### PR TITLE
BE nest.js + typescript + Express server setup

### PR DESCRIPTION
**Pull request by shnawaz844**

### CODE REVIEW

File: src/app.module.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { Module } from '@nestjs/common';
+import { AppService } from './app.service';
+import { TypeOrmModule } from '@nestjs/typeorm';
```
Comment (coder-faraz) [inline src/app.module.ts:2]: fix import order and sort each section alphabetically and separate each section with empty line

File: src/app.module.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import { UsersController } from './users/users.controller';
+import { TicketsController } from './tickets/tickets.controller';
+import { User } from './users/entities/user.entity';
+import { Ticket } from './tickets/entities/ticket.entity';
+import { Activity } from './ticket_activities/entities/ticket_activity.entity';
+
+@Module({
+  imports: [
+    TypeOrmModule.forRoot({
+      type: 'mysql',
```
Comment (coder-faraz) [inline src/app.module.ts]: this will come from env file

File: src/app.module.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      host: 'localhost',
+      port: 3306,
+      username: 'root',
+      password: '',
+      database: 'f2-fintech',
+      entities: [User, Ticket, Activity],
```
Comment (coder-faraz) [inline src/app.module.ts]: rename table name to TicketActivity

File: src/app.module.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      synchronize: true,
+    }),
+    UsersModule,
+    TicketsModule,
+    TicketActivitiesModule,
+  ],
+  controllers: [AppController, UsersController, TicketsController],
```
Comment (coder-faraz) [inline src/app.module.ts]: add TicketActivityController here

File: src/ticket_activities/dto/create-ticket_activity.dto.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { IsString, IsOptional, Length } from 'class-validator';
```
Comment (coder-faraz) [inline src/ticket_activities/dto/create-ticket_activity.dto.ts]: rename this file to underscore case, example create_ticket_activity

File: src/ticket_activities/dto/update-ticket_activity.dto.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { IsString, IsOptional, Length } from 'class-validator';
```
Comment (coder-faraz) [inline src/ticket_activities/dto/update-ticket_activity.dto.ts]: import the dto from create and use in this file, example - 
export class UpdateTicketActivityDto extends PartialType(CreateTicketActivityDto) {}

File: src/ticket_activities/ticket_activities.controller.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  Body,
+  Get,
+  Param,
+  Patch,
+  Delete,
+} from '@nestjs/common';
+import { ActivitiesService } from './ticket_activities.service';
+import { CreateTicketActivityDto } from './dto/create-ticket_activity.dto';
+
+@Controller('activities')
```
Comment (coder-faraz) [inline src/ticket_activities/ticket_activities.controller.ts]: rename activities to ticket_activities

File: src/ticket_activities/ticket_activities.controller.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Patch(':id')
+  update(
+    @Param('id') id: string,
+    @Body() updateTicketActivityDto: CreateTicketActivityDto,
+  ) {
+    return this.activitiesService.update(+id, updateTicketActivityDto);
+  }
+
+  @Delete(':id')
```
Comment (coder-faraz) [inline src/ticket_activities/ticket_activities.controller.ts]: remove the delete function

File: src/ticket_activities/ticket_activities.service.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    id: number,
+    updateTicketActivityDto: UpdateTicketActivityDto,
+  ): Promise<Activity> {
+    const activity = await this.findOne(id);
+    Object.assign(activity, updateTicketActivityDto, { updatedAt: new Date() });
+    return await this.activityRepository.save(activity);
+  }
+
+  // Remove an activity by ID
+  async remove(id: number): Promise<void> {
```
Comment (coder-faraz) [inline src/ticket_activities/ticket_activities.service.ts]: remove this function

File: src/ticket_activities/entities/ticket_activity.entity.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
+
+@Entity('activities')
```
Comment (coder-faraz) [inline src/ticket_activities/entities/ticket_activity.entity.ts]: rename activities to ticket_activities

File: src/tickets/tickets.controller.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  findOne(@Param('id') id: string) {
+    return this.ticketsService.findOne(+id);
+  }
+
+  @Patch(':id')
+  update(@Param('id') id: string, @Body() updateTicketDto: UpdateTicketDto) {
+    return this.ticketsService.update(+id, updateTicketDto);
+  }
+
+  @Delete(':id')
```
Comment (coder-faraz) [inline src/tickets/tickets.controller.ts]: remove delete function

File: src/tickets/tickets.service.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    return ticket;
+  }
+
+  async update(id: number, updateTicketDto: UpdateTicketDto): Promise<Ticket> {
+    const ticket = await this.findOne(id);
+    Object.assign(ticket, updateTicketDto, { updatedAt: new Date() });
+    return await this.ticketRepository.save(ticket);
+  }
+
+  async remove(id: number): Promise<void> {
```
Comment (coder-faraz) [inline src/tickets/tickets.service.ts]: remove this function

File: src/users/dto/create-user.dto.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  IsString,
+  IsEmail,
+  IsNumber,
+  IsEnum,
+  IsOptional,
+  Length,
+} from 'class-validator';
+
+// Define enums directly in this file
+export enum Gender {
```
Comment (coder-faraz) [inline src/users/dto/create-user.dto.ts]: import from entity

File: src/users/dto/create-user.dto.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+  MALE = 'male',
+  FEMALE = 'female',
+  OTHER = 'other',
+}
+
+export enum Status {
```
Comment (coder-faraz) [inline src/users/dto/create-user.dto.ts]: import from entity

File: src/users/entities/user.entity.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  @Column({ length: 100 })
+  username: string;
+
+  @Column()
+  password: string;
+
+  @Column({ unique: true })
+  email: string;
+
+  @Column({ length: 100 }) // Use 'bigint' for large numbers
```
Comment (coder-faraz) [inline src/users/entities/user.entity.ts]: length should be 15

File: src/users/entities/user.entity.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+  contact: string;
```
Comment (coder-faraz) [inline src/users/entities/user.entity.ts]: change contact: number;

File: src/users/users.controller.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  findOne(@Param('id') id: number) {
+    return this.usersService.findOne(+id);
+  }
+
+  @Patch(':id')
+  update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto) {
+    return this.usersService.update(+id, updateUserDto);
+  }
+
+  @Delete(':id')
```
Comment (coder-faraz) [inline src/users/users.controller.ts]: remove delete function

File: src/users/users.service.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+      where: { id }, // This is shorthand and should work as well
+    });
+  }
+
+  async update(id: number, updateUserDto: UpdateUserDto): Promise<User> {
+    await this.userRepository.update(id, updateUserDto);
+    return this.findOne(id);
+  }
+
+  async remove(id: number): Promise<void> {
```
Comment (coder-faraz) [inline src/users/users.service.ts]: remove delete function

File: src/users/users.service.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+  async create(createUserDto: CreateUserDto): Promise<User> {
+    const newUser = this.userRepository.create(createUserDto);
+    return await this.userRepository.save(newUser);
+  }
+
+  async findAll(): Promise<User[]> {
+    return await this.userRepository.find();
+  }
+
+  async findOne(id: number): Promise<User> {
```
Comment (coder-faraz) [inline src/users/users.service.ts:59]: make this function same as 
`  async findOne(id: number): Promise<Ticket> {
    const ticket = await this.ticketRepository.findOne({ where: { id } });
    if (!ticket) {
      throw new NotFoundException(`Ticket with ID ${id} not found`);
    }
    return ticket;
  }`

File: src/users/users.controller.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
+import {
```
Comment (coder-faraz) [inline src/users/users.controller.ts]: create spec (test) file for controller

File: src/users/users.service.ts [hunk 21]
```diff
@@ -0,0 +1,38 @@
+import { Injectable } from '@nestjs/common';
```
Comment (coder-faraz) [inline src/users/users.service.ts]: create spec (test) file for service

**Review by coder-faraz — CHANGES_REQUESTED** [review]
Create utils folder/file and create a formatResponse() function and use this function in every service before returning the response

File: src/app.module.ts (hunk 1 above)
Comment (shnawaz844) [reply src/app.module.ts:2]: done

File: src/app.module.ts (hunk 3 above)
Comment (shnawaz844) [reply src/app.module.ts]: Done

File: src/app.module.ts (hunk 4 above)
Comment (shnawaz844) [reply src/app.module.ts]: done

File: src/ticket_activities/dto/create-ticket_activity.dto.ts (hunk 5 above)
Comment (shnawaz844) [reply src/ticket_activities/dto/create-ticket_activity.dto.ts]: done

File: src/ticket_activities/dto/update-ticket_activity.dto.ts (hunk 6 above)
Comment (shnawaz844) [reply src/ticket_activities/dto/update-ticket_activity.dto.ts]: done

File: src/ticket_activities/entities/ticket_activity.entity.ts (hunk 10 above)
Comment (shnawaz844) [reply src/ticket_activities/entities/ticket_activity.entity.ts]: done

File: src/ticket_activities/ticket_activities.controller.ts (hunk 7 above)
Comment (shnawaz844) [reply src/ticket_activities/ticket_activities.controller.ts]: done

File: src/ticket_activities/ticket_activities.controller.ts (hunk 8 above)
Comment (shnawaz844) [reply src/ticket_activities/ticket_activities.controller.ts]: done

File: src/ticket_activities/ticket_activities.service.ts (hunk 9 above)
Comment (shnawaz844) [reply src/ticket_activities/ticket_activities.service.ts]: done

File: src/tickets/tickets.controller.ts (hunk 11 above)
Comment (shnawaz844) [reply src/tickets/tickets.controller.ts]: done

File: src/tickets/tickets.service.ts (hunk 12 above)
Comment (shnawaz844) [reply src/tickets/tickets.service.ts]: done

File: src/users/dto/create-user.dto.ts (hunk 13 above)
Comment (shnawaz844) [reply src/users/dto/create-user.dto.ts]: done

File: src/users/dto/create-user.dto.ts (hunk 14 above)
Comment (shnawaz844) [reply src/users/dto/create-user.dto.ts]: done

File: src/users/entities/user.entity.ts (hunk 15 above)
Comment (shnawaz844) [reply src/users/entities/user.entity.ts]: done

File: src/users/entities/user.entity.ts (hunk 16 above)
Comment (shnawaz844) [reply src/users/entities/user.entity.ts]: done

File: src/users/users.controller.ts (hunk 17 above)
Comment (shnawaz844) [reply src/users/users.controller.ts]: done

File: src/tickets/entities/tickets.controller.spec.ts [hunk 22]
```diff
@@ -0,0 +1,20 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (coder-faraz) [inline src/tickets/entities/tickets.controller.spec.ts]: move this file outside the entities folder

**Review by coder-faraz — CHANGES_REQUESTED** [review]
test cases of user controller and user service are missing

File: src/users/users.service.ts (hunk 18 above)
Comment (shnawaz844) [reply src/users/users.service.ts]: done

File: src/users/users.controller.ts (hunk 20 above)
Comment (shnawaz844) [reply src/users/users.controller.ts]: done

File: src/users/users.service.ts (hunk 21 above)
Comment (shnawaz844) [reply src/users/users.service.ts]: done

File: src/tickets/entities/tickets.controller.spec.ts (hunk 22 above)
Comment (shnawaz844) [reply src/tickets/entities/tickets.controller.spec.ts]: done

**Review by coder-faraz — APPROVED** [review]
good

File: src/app.module.ts (hunk 2 above)
Comment (shnawaz844) [reply src/app.module.ts]: done

**Review by coder-faraz — APPROVED** [review]
done

**Pull request merged**